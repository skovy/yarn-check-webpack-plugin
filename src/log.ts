import chalk from "chalk";

export const log = {
  success: (message: string) => console.log(chalk.green(message)),
  warn: (message: string) => console.log(chalk.yellow(message)),
  error: (message: string) => console.log(chalk.red(message))
};