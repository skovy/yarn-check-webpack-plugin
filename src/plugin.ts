import webpack from "webpack";

import { run } from ".";

class YarnCheck implements webpack.Plugin {
  apply(compiler: webpack.Compiler) {
    compiler.hooks.afterCompile.tapAsync("YarnCheck", async (_, callback) => {
      await run();
      callback();
    });
  }
}

export { YarnCheck };
