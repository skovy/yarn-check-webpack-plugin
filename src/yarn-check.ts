import { spawn } from "child_process";
import chalk from "chalk";

import { log } from "./log";

const PACKAGE_NOT_INSTALLED = /\"(.*)\" not installed/;
const PACKAGE_WRONG_VERSION = /\"(.*)\" is wrong version: expected \"(.*)\", got \"(.*)\"/;

export interface PluginOptions {
  /**
   * The root directory to run the commands. This should be the directory that
   * contains `package.json`, `yarn.lock` and `node_modules`. By default, it will
   * run in the current directory. This option only needs to be set if this
   * plugin is being run in a different directory.
   */
  rootDirectory?: string;

  /**
   * Ignore any missing or wrong version packages from this plugins warnings.
   * This argument
   */
  exclude?: RegExp;
}

interface Package {
  name: string;
  version?: {
    expected: string;
    got: string;
  };
}

const shouldPreventWarning = (
  pkg: string,
  exclude: RegExp | undefined
): boolean => {
  if (!exclude) {
    return false;
  }

  return exclude.test(pkg);
};

const runCheckVerifyTree = async ({
  rootDirectory,
  exclude
}: PluginOptions): Promise<{
  packagesNotInstalled: Package[];
  packagesWrongVersion: Package[];
}> => {
  return new Promise(resolve => {
    const packagesNotInstalled: Package[] = [];
    const packagesWrongVersion: Package[] = [];

    const yarnCheckVerifyTree = spawn(`yarn`, [`check`, `--verify-tree`], {
      cwd: rootDirectory,
      env: process.env
    });

    yarnCheckVerifyTree.stderr.on("data", data => {
      const err = data.toString();

      const packageNotInstalled = err.match(PACKAGE_NOT_INSTALLED);
      if (packageNotInstalled) {
        const name = packageNotInstalled[1];

        if (shouldPreventWarning(name, exclude)) return;

        packagesNotInstalled.push({ name });
      }

      const packageWrongVersion = err.match(PACKAGE_WRONG_VERSION);
      if (packageWrongVersion) {
        const [, name, expected, got] = packageWrongVersion;

        if (shouldPreventWarning(name, exclude)) return;

        packagesWrongVersion.push({ name, version: { expected, got } });
      }
    });

    yarnCheckVerifyTree.on("exit", () => {
      resolve({ packagesNotInstalled, packagesWrongVersion });
    });
  });
};

/**
 * Run the yarn check command, parse any missing or incorrect packages, log
 * the issues and any other helpful prompts.
 * 
 * @param options an object to configure how the yarn check is ran
 * 
 * @returns a promise that resolves to a boolean. `true` if the method ran 
 * successfully with no issues. `false` if there was one more missing or 
 * incorrect packages.
 */
export const run = async (options: PluginOptions = {}): Promise<boolean> => {
  const {
    packagesNotInstalled,
    packagesWrongVersion
  } = await runCheckVerifyTree(options);

  if (!packagesNotInstalled.length && !packagesWrongVersion.length) {
    log.success(`All packages installed and up to date.`);
    return true;
  }

  if (packagesNotInstalled.length) {
    log.warn(`Missing packages:`);
    packagesNotInstalled.map(pkg => log.warn(`  - ${pkg.name}`));
  }

  if (packagesWrongVersion.length) {
    log.warn(`Wrong packages:`);
    packagesWrongVersion.map(pkg =>
      log.warn(
        `  - ${pkg.name} (expected ${pkg.version!.expected} but got ${
          pkg.version!.got
        })`
      )
    );
  }

  log.error(
    `Please run ${chalk.bold(`\`yarn install --check-files\``)} to update.`
  );

  return false;
};
