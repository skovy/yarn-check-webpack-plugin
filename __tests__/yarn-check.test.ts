import { fixtures } from "./fixtures";

import { run } from "../src/yarn-check";
import { log } from "../src/log";

describe("yarn-check", () => {
  let successSpy: jest.SpyInstance;
  let warnSpy: jest.SpyInstance;
  let errorSpy: jest.SpyInstance;

  beforeEach(() => {
    successSpy = jest.spyOn(log, "success").mockImplementation();
    warnSpy = jest.spyOn(log, "warn").mockImplementation();
    errorSpy = jest.spyOn(log, "error").mockImplementation();
  });

  afterEach(() => {
    successSpy.mockRestore();
    warnSpy.mockRestore();
    errorSpy.mockRestore();
  });

  describe("a project that is up to date", () => {
    it("lists missing packages, wrong versions and a prompt to install", async () => {
      await run({ rootDirectory: fixtures.upToDateProject });

      expect(successSpy.mock.calls.join("\n")).toMatchSnapshot();
      expect(warnSpy.mock.calls.join("\n")).toMatchSnapshot();
      expect(errorSpy.mock.calls.join("\n")).toMatchSnapshot();
    });
  });

  describe("a project with a missing package and wrong package", () => {
    it("lists missing packages, wrong versions and a prompt to install", async () => {
      await run({ rootDirectory: fixtures.missingAndWrongPackage });

      expect(successSpy.mock.calls.join("\n")).toMatchSnapshot();
      expect(warnSpy.mock.calls.join("\n")).toMatchSnapshot();
      expect(errorSpy.mock.calls.join("\n")).toMatchSnapshot();
    });
  });

  describe("options", () => {
    describe("exclude", () => {
      it("does not warn about missing packages that are excluded", async () => {
        await run({
          rootDirectory: fixtures.missingAndWrongPackage,
          exclude: /under/
        });

        expect(warnSpy.mock.calls.join("")).toContain("classnames");
        expect(warnSpy.mock.calls.join("")).not.toContain("underscore");
      });

      it("does not warn about wrong version packages that are excluded", async () => {
        await run({
          rootDirectory: fixtures.missingAndWrongPackage,
          exclude: /names/
        });

        expect(warnSpy.mock.calls.join("")).toContain("underscore");
        expect(warnSpy.mock.calls.join("")).not.toContain("classnames");
      });
    });
  });
});
