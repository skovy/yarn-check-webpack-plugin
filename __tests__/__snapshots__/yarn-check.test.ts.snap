// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yarn-check a project that is up to date prints a success message 1`] = `"All packages installed and up to date."`;

exports[`yarn-check a project that is up to date prints a success message 2`] = `""`;

exports[`yarn-check a project that is up to date prints a success message 3`] = `""`;

exports[`yarn-check a project with a missing package and wrong package lists missing packages, wrong versions and a prompt to install 1`] = `""`;

exports[`yarn-check a project with a missing package and wrong package lists missing packages, wrong versions and a prompt to install 2`] = `
"Missing packages:
  - underscore
  - object-assign
Wrong packages:
  - classnames (expected ^2.2.6 but got 3.2.6)"
`;

exports[`yarn-check a project with a missing package and wrong package lists missing packages, wrong versions and a prompt to install 3`] = `"Please run [1m\`yarn install --check-files\`[22m to update."`;
